Add Azure OpenAI Support

thinkthinking opened the pull request:
Add Azure OpenAI Support

msabramo commented on README.md:
```diff
@@ -44,21 +44,25 @@ On a Linux/MacOS system it is defined by the $XDG_CONFIG_HOME variable (check it
-|`gpt-4-32k`|0.06|0.12|
-|`gpt-4-32k-0613`|0.06|0.12|
-|`gpt-4-1106-preview`|0.01|0.03|
+| Name                 | Pricing (input token) | Pricing(output token) |
+| -------------------- | --------------------- | --------------------- |
+| `gpt-3.5-turbo`      | 0.001                 | 0.002                 |
+| `gpt-3.5-turbo-1106` | 0.001                 | 0.002                 |
+| `gpt-3.5-turbo-0613` | 0.001                 | 0.002                 |
+| `gpt-3.5-turbo-16k`  | 0.001                 | 0.002                 |
+| `gpt-4`              | 0.03                  | 0.06                  |
+| `gpt-4-0613`         | 0.03                  | 0.06                  |
+| `gpt-4-32k`          | 0.06                  | 0.12                  |
+| `gpt-4-32k-0613`     | 0.06                  | 0.12                  |
+| `gpt-4-1106-preview` | 0.01                  | 0.03                  |
```
Sooo much nicer! A nice change in its own right!

msabramo commented on src/chatgpt.py:
```diff
@@ -22,21 +22,25 @@
-BASE_ENDPOINT = "https://api.openai.com/v1"
-ENV_VAR = "OPENAI_API_KEY"
 
+# azure price is not accurate, it deponds on your subscription
```
deponds => depends

msabramo commented on src/chatgpt.py:
```diff
@@ -181,10 +188,7 @@ def display_expense(model: str) -> None:
-        f"\nTotal tokens used: [green bold]{prompt_tokens + completion_tokens}",
-        extra={"highlighter": None},
-    )
+    logger.info(f"\nTotal tokens used: [green bold]{prompt_tokens + completion_tokens}",extra={"highlighter": None},)
```
Old version looks nicer here IMHO

msabramo commented:
Overall this is really nice and I tested it out with an Azure OpenAI instance I have access to and it worked well. Thanks for sharing this!

If I were the maintainer, I might suggest separating some of the non-Azure stuff like `test_connection` and formatting changes into separate PRs, but I'll leave them up to @marcolardera.

msabramo commented on src/chatgpt.py:
```diff
@@ -302,10 +365,22 @@ def start_prompt(
         raise KeyboardInterrupt
 
     messages.append({"role": "user", "content": message})
+    if config["supplier"] == "azure":
+        api_key = config["AZURE_OPENAI_API_KEY"]
+        # Base body parameters
+        model = config["AZURE_OPENAI_DEPLOYMENT_NAME"]
+        api_version = config['AZURE_OPENAI_API_VERSION']
+        base_endpoint = config["AZURE_OPENAI_ENDPOINT"]
+    elif config["supplier"] == "openai":
+        api_key = config["OPENAI_API_KEY"]
+        model = config["OPENAI_MODEL"]
+        base_endpoint = config["OPENAI_ENDPOINT"]
+    else:
+        raise ValueError("Supplier must be either 'azure' or 'openai'")
```
I wonder if it makes sense at this point to create a `Supplier` class and then have `OpenAISupplier` and `AzureOpenAISupplier` subclasses? Otherwise this could get long and hard to maintain over time.

msabramo commented on src/chatgpt.py:
```diff
@@ -316,9 +391,22 @@ def start_prompt(
-        r = requests.post(
-            f"{BASE_ENDPOINT}/chat/completions", headers=headers, json=body
-        )
+        if config["supplier"] == "azure":
+            headers = {
+                "Content-Type": "application/json",
+                "api-key": api_key,
+            }
+            r = requests.post(
+                f"{base_endpoint}/openai/deployments/{model}/chat/completions?api-version={api_version}", headers=headers, json=body
+            )
+        elif config["supplier"] == "openai":
+            headers = {
+                "Content-Type": "application/json",
+                "Authorization": f"Bearer {api_key}",
+            }
+            r = requests.post(
+                f"{base_endpoint}/chat/completions", headers=headers, json=body
+            )
```
More switching on the `supplier` - makes me think that having classes as mentioned before will be nice.

msabramo commented on src/chatgpt.py:
```diff
@@ -487,14 +576,25 @@ def main(
 
     copyable_blocks = {} if config["easy_copy"] else None
-
+    if config["supplier"] == "azure":
+        model = config["AZURE_OPENAI_DEPLOYMENT_NAME"]
+    elif config["supplier"] == "openai":
+        model = config["OPENAI_MODEL"]
```
More switching on the `supplier` - makes me think that having classes as mentioned before will be nice.

marcolardera replied:
Your consideration is correct. In general all the code is becoming a little bit messy after many addition and it needs a general refactoring. I will probably do it during Christmas vacation.

marcolardera commented:
Hello, I like this feature but I needed to make some adjustments. In particular:

- Since the config parameters names were modified, even the old ones, this was a breaking change (people would have had the configuration broken after an update to this version). And I don't like that at all. So I changed the config names in a way that the OpenAI parameters remains the same as before.
- I added the support for the --key and --model options in case the supplier is set to Azure (more ugly if-else...yayy)
- I removed the test_connection function. I think that it adds more complexity without much benefit. It also increases the expenses by a tiny bit.

@thinkthinking @msabramo Can I ask you to pull my last commit on this branch and confirm that using Azure as a supplier still works correctly? After that I can merge into main. Thank you very much!

AsafMah commented:
Works for me!

thinkthinking commented:
> Hello, I like this feature but I needed to make some adjustments. In particular:
> 
> * Since the config parameters names were modified, even the old ones, this was a breaking change (people would have had the configuration broken after an update to this version). And I don't like that at all. So I changed the config names in a way that the OpenAI parameters remains the same as before.
> * I added the support for the --key and --model options in case the supplier is set to Azure (more ugly if-else...yayy)
> * I removed the test_connection function. I think that it adds more complexity without much benefit. It also increases the expenses by a tiny bit.
> 
> @thinkthinking @msabramo Can I ask you to pull my last commit on this branch and confirm that using Azure as a supplier still works correctly? After that I can merge into main. Thank you very much!

Hello,

I've received your message and will pull the latest commit to check that Azure is still functioning correctly as a supplier with our configurations. I will run the necessary tests and provide you with an update once completed.

Thank you for making sure we move forward with backward compatibility and for streamlining our codebase. I'll get back to you with the results as soon as possible.

Best,
thinkthinking

marcolardera commented:
After a bit of testing, I feel it could be merged.

Thank you all!
